### PR TITLE
Java: add websocket reads as remote flow source.

### PR DESCRIPTION
**Pull request by porcupineyhairs**
Currently, JAX-WS reads are considered as untrusted. However, `java.net.http.WebSocket` reads are not marked as such. This PR adds support for the same.

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import java
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll:1]: This needs a file-level qldoc along the lines of `/** Provides classes for reasoning about ...`.

File: java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+/**Models `java.net.http.Websocket` package's listener interface. */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll:3]:
```suggestion
/** The `java.net.http.Websocket.Listener` interface. */
```

File: java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+class WebsocketListener extends Interface {
+  WebsocketListener() { this.hasQualifiedName("java.net.http", "WebSocket$Listener") }
+}
+
+/**Models `java.net.http.Websocket.Listener` interface's `onText` method */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll:8]:
```suggestion
/** The method `onText` on a type that implements the `java.net.http.Websocket.Listener` interface. */
```

File: java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+class WebsocketOnText extends Method {
+  WebsocketOnText() {
+    exists(WebsocketListener l |
+      this.getDeclaringType().extendsOrImplements(l) and
+      // onText​(WebSocket webSocket, CharSequence data, boolean last)
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/javase/WebSocket.qll:13]:
```suggestion
      // onText(WebSocket webSocket, CharSequence data, boolean last)
```
Remove U+200B character.

File: java/ql/test/query-tests/security/CWE-079/semmle/tests/WebsocketXss.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+public class WebsocketXss {
+    public static void main(String[] args) throws Exception {
+        WebSocket.Listener listener = new WebSocket.Listener() {
+            public CompletionStage<?> onText(WebSocket webSocket, CharSequence message, boolean last) {
+                try {
+                    HttpClient client = HttpClient.newBuilder().build();
+                    CompletableFuture<WebSocket> ws = client.newWebSocketBuilder()
+                            .buildAsync(URI.create("ws://websocket.example.com"), null);
+                    ws.get().sendText​(message, false);
```
Comment (aschackmull) [inline java/ql/test/query-tests/security/CWE-079/semmle/tests/WebsocketXss.java:15]:
```suggestion
                    ws.get().sendText(message, false);
```
Remove U+200B character.

**Comment by porcupineyhairs** [comment]
@aschackmull This is dependent on https://github.com/github/securitylab/discussions/142 

The WebSocket package was only available starting Java 11, CodeQL extractor by default uses Java 8. Hence, the tests don't compile and run. The test code does not depend on any third party libraries so there's nothing I need to stub here. 

As for the positive result on LGTM, I dont know why only that one was detected. Github's fuzzy search returns about 4000+ results for `websocket onText` So we should be seeing more results on LGTM provided the query runs correctly.

**Comment by intrigus-lgtm** [comment]
> As for the positive result on LGTM, I dont know why only that one was detected. Github's [fuzzy search](https://github.com/search?q=onText+websocket+language%3Ajava&type=Code) returns about 4000+ results for `websocket onText` So we should be seeing more results on LGTM provided the query runs correctly.

Websocket are pretty new.
They have been added with Java 11.

Most of Github's fuzzy search results are in forks of the jdk sources.

**Comment by aschackmull** [comment]
For the test you can add a file called `options` next to the test containing the following single line:
```
//semmle-extractor-options: --javac-args -source 11 -target 11
```
That makes the test run with Java 11 instead of 8.

**Comment by porcupineyhairs** [comment]
@aschackmull Unfortunately, that does not work. You may see in the `javac-extractor-*.log`, the command generated is 
```
sun.java.command=com.semmle.extractor.java.JavaExtractor --strict-javac-errors --javac-args -source 8 -target 8 -source 11 -target 11 -cp <Redacted>/java/ql/test/query-tests/security/CWE-079/semmle/tests/../../../../../stubs/servlet-api-2.4/ -encoding UTF-8 WebsocketXss.java XSS.java
```

The `-source 8` and `-target  8` flags are added regardless of whether they are specified explicitly. This looks like a bug in the extractor to me. WDYT?

**Comment by aschackmull** [comment]
I've done some debugging and found the reason why this qltest fails to compile.  For the sake of reproducibility we run all the qltests in a fixed and fairly minimal JDK, and this doesn't include the `java.net.http` module. I've tried to locally upgrade the qltest JDK to include the `java.net.http` module, and then the test compiles.

For this PR, I think the best way forward is to commit a completely commented-out version of the `WebsocketXss.java` test file - then I'll test it locally and work to enable it once the tooling supports the `java.net.http` module.

Btw., when running the test locally I found that it didn't produce any new results - I guess the XSS sink is also missing. Presumably `WebSocket.sendText` should be considered an XSS sink? (I'm not familiar with WebSockets, so please correct me if I'm wrong here). Do you want to fix this here, or should we address this in a separate PR?

**Comment by porcupineyhairs** [comment]
@aschackmull Yes, the `sendText` function is a potential sink. However. it can be used to send almost anything even json responses. By marking all `sendText` calls as sinks, we may significantly increase the number of false postiives in our results.
One more important point to note is that the sent message may be sanitized on the client side too. So we have no way of knowing for sure if a true positive is actually a true positive. 

For my corresponding Golang [PR](https://github.com/github/codeql-go/pull/109), I had included the send calls as sinks but those were rejected by the team and hence had to be removed. See this comment [here ](https://github.com/github/codeql-go/pull/109#issuecomment-628076498) for the context.

That's the reason I didn't include them in this one. Please let me know if you think otherwise. I would be more than happy to add them here.

I have comment out the `.java` file and rebase it on the latest master. Sorry for the force push. The commit somehow got messed up during merge.

**Comment by aschackmull** [comment]
Alright, this looks almost ready to merge, then. Could I just ask you to remove the U+200B character in `WebsocketXss.java`? It's failing our ascii check. It's an invisible non-breaking space that comes from copy-pasting.

**Comment by porcupineyhairs** [comment]
@aschackmull Done :+1:

**Review by aschackmull — APPROVED** [review]

**Pull request merged**